lib/adm.js

```diff
@@ -4097,6 +4121,106 @@ maAdm.prototype.generatePlan = function (opts, callback)
 	setTimeout(callback, 0);
 };
 
+/*
+ * Determine whether the plan that was generated adheres to any restrictions
+ * that are in place.
+ * For now the only check implemented is to ensure that all images for
+ * deployment are available on the specific imgapi channel set by
+ * 'sdcadm channel set' on the 'remote_imgapi' set in etc/config.json.
+ *
+ * Options:
+ * - skip_verify_channel: if true, we avoid verifying that each image to be
+ *   provisioned is present on the defined SDC updates.joyent.com channel.
+ */
+maAdm.prototype.verifyPlan = function (opts, cb)
+{
+	assert.ok(this.ma_plan !== null);
+	assertplus.object(opts, 'opts');
+	assertplus.func(cb, 'cb');
+	assertplus.bool(opts.skip_verify_channel, 'opts.skip_verify_channel');
+
+	var self = this;
+
+	if (opts.skip_verify_channel) {
+		return (cb());
+	}
+	/*
+	 * If we're doing channel verification, go through our plan,
+	 * checking that each new image exists on that channel. To do this,
+	 * find the set of new image uuids per-service across all compute nodes
+	 * in the plan.
+	 */
+	// We don't have a Set() object, so instead use unique Object keys
+	var services_by_image = {};
+	for (var svcname in this.ma_plan) {
+		for (var cnid in this.ma_plan[svcname]) {
+			for (var i = 0;
+			    i < this.ma_plan[svcname][cnid].length;
+			    i++) {
+				var plan_op = this.ma_plan[svcname][cnid][i];
+				var new_image = null;
+				if (plan_op.action === 'provision') {
+					new_image = plan_op.IMAGE;
+				} else if (plan_op.action === 'reprovision') {
+					new_image = plan_op.new_image;
+				}
+				if (new_image) {
+					services_by_image[new_image] = svcname;
+				}
+			}
+		}
+	}
+
+	var imgapi_opts = {'channel': self.ma_channel};
+
+	// An array of strings, later used to form an error message.
+	var imagesNotInTheChannel = [];
+	// process each of the images in series
+	vasync.forEachParallel({
+	    inputs: Object.keys(services_by_image),
+	    func: function channelCheckOneImage(imageUuid, nextImage) {
+			self.ma_log.debug(
+			    sprintf(
+				    'looking up image %s for service %s ' +
+				    'on channel %s',
+				    imageUuid, services_by_image[imageUuid],
+				    self.ma_channel));
+			    self.ma_sdc.REMOTE_IMGAPI.getImage(
+				    imageUuid, imgapi_opts,
+			    function getImage(err, image) {
+				if (err) {
+					// note that we don't invoke the
+					// err callback since vasync will
+					// then report only
+					// 'the first of n errors'
+					var missing = sprintf('%s (%s)',
+					    imageUuid,
+					    services_by_image[imageUuid]);
+					self.ma_log.debug(
+					    'missing image ' + missing);
+					imagesNotInTheChannel.push(missing);
+					nextImage();
+				} else {
+					nextImage();
+				}
+			});
+		}
+	}, function finish(err) {
+		if (err) {
+			cb(err);
+		} else if (imagesNotInTheChannel.length > 0) {
+			cb(new VError(
+				'The following images were not found on ' +
+				'the \'' + self.ma_channel + '\' channel ' +
```

Comment:
VError supports sprintf syntax if you *like*, so one doesn't have to do the `'...' + myVar + '...'` thing. Eitrher way is fine.

Personally, given that we use single-quotes for string delims, I tend to use *double-quotes* in cases like this. Double quotes in error messages and log messages are just fine (arguably even better when quoting contractions), and it makes code like this tidier:

```
'the "%s" channel', self.ma_channel
```

Reply:
That looks much nicer. I'll change that here, and look for other occurrences within the scope of my changes.